### PR TITLE
Enable running as non root

### LINKED ISSUE #273: Building custom image from caddy:2.6.2 failed
Dear caddy-docker maintainer,

I've tried to build a custom image from your Alpine Dockerfile (caddy-docker/2.6/alpine/Dockerfile) with the following command:

`docker build --tag <custom-tag> .`

During the building process I encountered the following error message:

```
Step 6/18 : RUN [ ! -e /etc/nsswitch.conf ] && echo 'hosts: files dns' > /etc/nsswitch.conf
 ---> Running in 3762efc54156
The command '/bin/sh -c [ ! -e /etc/nsswitch.conf ] && echo 'hosts: files dns' > /etc/nsswitch.conf' returned a non-zero code: 1
```

I guess there might be a mistake since the test command checks if `/etc/nsswitch.conf` is **not** available and in case write to it. You probably mean to check whether `/etc/nsswitch.conf` is indeed available. So no exclamation mark in test command. I guess the  RUN statement was supposed to look like the following:

`RUN [ -e /etc/nsswitch.conf ] && echo 'hosts: files dns' > /etc/nsswitch.conf`

Correct me if I'm wrong or missing something.

Best regards!

### PR DESCRIPTION
**Pull request by abjugard**
Addresses part of #104 

Fixes #273

### CODE REVIEW

**Comment by francislavoie** [comment]
I don't understand what CI is doing :thinking: it's erroring out on the nsswitch.conf thing, but we removed that.

This will need @hairyhenderson's eyes on it I think.

I opened my own PR #275 and I guess this one precludes that one since it makes the same fix.

**Comment by abjugard** [comment]
(No change, just added the #273 issue tag to the relevant commit)

**Comment by francislavoie** [comment]
Oh, really? It needs to be set on the binary before copying it over? Wow, that's really annoying. I thought `setcap` would write a rule to the existing container (i.e. the final stage) that would persist when the binary changes at the same path. I guess that's not how it works? I've never dug into how `setcap` really works.

I'd really hope not to have to add that to the recommended instructions. Is there anything we could do to make it smoother without needing to add that line?

An idea, we could add an option to `xcaddy` to try to run `setcap` on the binary after it's done building it. Would that work at all? Then we could use an env var to instruct `xcaddy` to do so, so the user doesn't see any difference in their own Dockerfile.

**Comment by abjugard** [comment]
Having `xcaddy` run setcap on the binary is enough yeah, and sounds like an elegant solution in my opinion!

**Comment by francislavoie** [comment]
> how do we instruct users that it's possible to do that, since that would still require changes to the Dockerfile?

We'd add it to here, maybe `XCADDY_SETCAP_NET=1` or something:

https://github.com/caddyserver/caddy-docker/blob/aca93aecdbd9ba4b4229fba1e6c7ee7b87d5cf7f/Dockerfile.builder.tmpl#L11

**Comment by jjlin** [comment]
I tried doing something similar for another project a couple of years ago, and there are some big caveats. I haven't checked whether they still apply, but I wouldn't be surprised at all if they do.

1. Capabilities are associated with an executable via extended file attributes. The `COPY` instruction doesn't copy extended attributes, so you can't do a multi-stage build where you set the capability in the build stage and then copy that out to a runtime stage.

2. If you set a capability via a separate `RUN` instruction, Docker seems to make another copy of the executable in a new layer (with an image where the bulk of the content is the executable, the image basically doubled in size). So it's probably best to build the executable and set the capability via a single `RUN`.

3. I've read that support for extended attributes may not be the same across all Linux platforms, though I don't have personal experience beyond the common ones. Since the Caddy image is built for some of the less common platforms, that might be something to verify.

This [blog post](https://wildwolf.name/multi-stage-docker-builds-and-xattrs/) is a couple of years old, but seems to cover most of what I observed, and more.

File: Dockerfile.builder.tmpl
Comment (francislavoie) [inline Dockerfile.builder.tmpl:5]: I think I'd like to keep this as still one-package-per-line, for readability's sake.

Comment (abjugard) [reply Dockerfile.builder.tmpl:5]: Should I change the main image one to match then? I just thought there should be a common style between them.

**Comment by francislavoie** [comment]
Thanks @jjlin, that's a useful article.

> 1. Capabilities are associated with an executable via extended file attributes. The COPY instruction doesn't copy extended attributes, so you can't do a multi-stage build where you set the capability in the build stage and then copy that out to a runtime stage.

That seems to have been true for the original Docker builder, but with Buildx/BuildKit, it is preserved. The article you linked mentions that. BuildKit is on by default now in latest versions of Docker, so it should work fine if users are keeping Docker up to date.

> 2. If you set a capability via a separate RUN instruction, Docker seems to make another copy of the executable in a new layer (with an image where the bulk of the content is the executable, the image basically doubled in size). So it's probably best to build the executable and set the capability via a single RUN.

Yep, that's a good point, and that's why I'd like for `xcaddy` to do it for us.

This PR adds `setcap` to the main `RUN` already for the main image, so that's already good :+1: 

> 3. I've read that Docker support for extended attributes may not be the same across all Linux platforms, though I don't have personal experience beyond the common ones. Since the Caddy image is built for some of the less common platforms, that might be something to verify.

:thinking: yeah, I don't have experience with this. We'll raise it with the Docker Library team when we push it there, they can tell us what's up.

File: Dockerfile.builder.tmpl
Comment (francislavoie) [reply Dockerfile.builder.tmpl:5]: I guess so.

Thanks!

Comment (abjugard) [reply Dockerfile.builder.tmpl]: No problem, done!

**Comment by abjugard** [comment]
> 🤔 yeah, I don't have experience with this. We'll raise it with the Docker Library team when we push it there, they can tell us what's up.

For what it's worth it works as we expect it to on my machine, documented my experience in https://github.com/caddyserver/xcaddy/issues/128

**Comment by francislavoie** [comment]
Oooh that's awesome, thanks for trying that and finding that new feature in Moby!

I think buildkit might only be enabled by default in new installs, not in upgraded ones. That's what I read in the docs I think.

File: 2.6/alpine/Dockerfile
Comment (hairyhenderson) [inline 2.6/alpine/Dockerfile:60]: I see a missing newline as the only change - I suspect your editor is misconfigured... Can you restore this? (same in the other files in this PR)

**Comment by hairyhenderson** [comment]
@abjugard can you revert the changes to `library/caddy` for now? I usually only update that separately after PRs have merged, to ensure the commits are correct for the main branch.

File: 2.6/alpine/Dockerfile
Comment (jjlin) [inline 2.6/alpine/Dockerfile:6]: Nitpick, but it would be nicer to have the package lists sorted here and elsewhere in this PR

Comment (francislavoie) [reply 2.6/alpine/Dockerfile:6]: Sure, I have no strong preference for ordering. I tend to leave them in the order I thought to add them, personally. But not for any particular reason.

Comment (abjugard) [reply 2.6/alpine/Dockerfile]: Alright, I've reordered the packages.

**Comment by abjugard** [comment]
> @abjugard can you revert the changes to `library/caddy` for now? I usually only update that separately after PRs have merged, to ensure the commits are correct for the main branch.

Done!

File: 2.6/alpine/Dockerfile
Comment (abjugard) [reply 2.6/alpine/Dockerfile:60]: Actually, there is an .editorconfig file in this repo, and these Dockerfiles do not adhere to its rules, but my editor does so I disabled some rules to keep my changes minimal, but I seem to have missed one.

Comment (abjugard) [reply 2.6/alpine/Dockerfile:60]: The rules say that files with extension .tmpl should have `insert_final_newline = false`, which results in the Dockerfiles also having that once generated, even though the .editorconfig would otherwise ensure those had a newline at the end.

Comment (abjugard) [reply 2.6/alpine/Dockerfile:60]: Let me know how you think we should proceed and I'll happily oblige, but it's hard to know exactly how you guys want styling considering you're not obeying your own configurations, and are using different styling in different templates but apparently explicitly prefer one over the other...

Comment (francislavoie) [reply 2.6/alpine/Dockerfile:60]: Sorry about that 😓 don't mean to cause friction!

Thanks for all the help, it's very much appreciated.

I'll look into making a PR for xcaddy today (hopefully).

**Comment by francislavoie** [comment]
Opened a PR in `xcaddy`: https://github.com/caddyserver/xcaddy/pull/129

File: 2.6/alpine/Dockerfile
Comment (abjugard) [reply 2.6/alpine/Dockerfile:60]: Haha it's no problem guys, just pointing it out 😛

Comment (abjugard) [reply 2.6/alpine/Dockerfile:60]: @hairyhenderson How do you think we should proceed? Should I update the .editorconfig according to your expectations? Should I "normalise" the repo to adhere to the .editorconfig? Should we delete the .editorconfig?

Comment (hairyhenderson) [reply 2.6/alpine/Dockerfile:60]: sorry, I missed these responses... let's delete the `.editorconfig` - IMO it's not useful in this repo.

Comment (hairyhenderson) [reply 2.6/alpine/Dockerfile:60]: ok, done in 6fa9eafebf128071ebdd129b459ecd8e960bd470

**Comment by hairyhenderson** [comment]
Do we need caddyserver/xcaddy#129 to merge first? otherwise this LGTM (except that this needs to be rebased, and the Dockerfiles regenerated now that the errant `.editorconfig` is gone)

**Comment by francislavoie** [comment]
Yeah, I think we should wait for xcaddy. This is only a half solution until we can make the builder also work correctly. Shouldn't take long, Matt was on vacation but will get back into thing soon.

**Comment by francislavoie** [comment]
Alright @abjugard, https://github.com/caddyserver/xcaddy/releases/tag/v0.3.2 is released! Got time to finish this up? :tada: 

Caddy v2.6.3 is gonna release sometime this week, would be great to get this in for that version.

**Comment by abjugard** [comment]
Absolutely @francislavoie. Want me to bump xcaddy in this PR then? There are some hashes there in the config, where are they taken from?

**Comment by abjugard** [comment]
How's that @francislavoie?

**Review by francislavoie — APPROVED** [review]
LGTM! I'll let Dave do the final review and merge

**Review by hairyhenderson — APPROVED** [review]
LGTM! Thanks for doing this @abjugard

**Pull request merged**
